### ISSUE TITLE
Add Kythe-CS, a web-based cross-referencing code search UI. [phab:D807]

### ISSUE THREAD

**Issue by kythe** [google/kythe#1101]
_Created by **fromberger** at 2016-04-03 09:12:14_: 

This code implements a web-based UI that allows users to search a corpus of
code using regular expressions and follow cross references between references,
definitions, derived classes and virtual overrides in a similar fashion to
websites such as http://cs.chromium.org/ . This is the code that currently
powers the LLVM code search website at http://llvm-cs.pcc.me.uk/ .

Depends on #1100

**Comment by kythe** [comment]
_**pcc** wrote at 2016-04-17 05:16:23_:

  - Add dump_cmdline.sh script
  - Collect definitions using both completion locations and definition locations
  - Clean up srcinfo dump formatting
  - Fix a typo that was causing us to drop derived classes

**Comment by kythe** [comment]
_**pcc** wrote at 2016-04-17 06:41:13_:

  - Add package documentation, add documentation for exported entities and fix other golint issues

**Comment by kythe** [comment]
_**pcc** wrote at 2016-04-17 06:55:59_:

  - Reword documentation in order to placate linter further

**Comment by kythe** [comment]
_**pcc** wrote at 2016-04-18 01:01:58_:

  - Use correct path to index command
  - Add a test for for-range loops

**Comment by kythe** [comment]
_**pcc** wrote at 2016-07-17 20:19:54_:

  - Allow names to not contain # character (for Java I believe)
  - "/" key focuses search box (based on code by Nico Weber)
  - Add check for empty path

**Comment by kythe** [comment]
_**pcc** wrote at 2016-12-04 11:21:06_:

  - Allow entering '/' in the search box (thanks Nico Weber!)
  - Update go rules

**Comment by kythe** [comment]
_**robinp** wrote at 2017-04-01 07:06:46_:

It's not clear why a separate server is needed for this - couldn't the Kythe crossref server (exposed by the http_server) be sufficient as backend?

**Comment by kythe** [comment]
_**pcc** wrote at 2017-04-03 05:58:07_:

>>! In (earlier comment), robinp wrote:
> It's not clear why a separate server is needed for this - couldn't the Kythe crossref server (exposed by the http_server) be sufficient as backend?

I believe that the xref server is more meant as a sample service than as something that is intended for production use. Also, the xref server requires an xref database, and building that database is not required (and somewhat counterproductive) for this particular application, as there is no need to perform queries at serving time if the serving data contains the answers to those queries in the form of static html pages. For this application, static html pages are pretty much all that is required for everything except search, for which we just need a few specialized indexes. With that as the serving data, the application is more efficient both on the server side and on the client side.

On the server side, the serving data can be much smaller than an entire xref database, which means that it can be served with fewer resources. For llvm-cs the serving data consumes about 5GB (at present) and the indexes fit in less than a gigabyte of memory. That allows the index to be served from a single g1-small GCE instance. (The index is built on a more powerful instance, but it is preemptible and it only needs to run while the index is being built.)

On the client side, static HTML pages allow for significantly faster load times, especially with larger source files. With pre-computed HTML, the client just needs to parse HTML, but with an xref service the client would need to query a server, parse the response, build HTML and parse that HTML (which is roughly what cs.chromium.org does I believe). You can see the difference for yourself by comparing these two pages, which are indexed versions of the current largest non-generated .cpp file in llvm.
https://cs.chromium.org/chromium/src/third_party/swiftshader/third_party/LLVM/lib/Target/X86/X86ISelLowering.cpp?type=cs&q=x86isellowering.cpp&l=9802
http://llvm-cs.pcc.me.uk/lib/Target/X86/X86ISelLowering.cpp

Chromium's version of the file is much older, so it isn't an entirely fair comparison, but cs.chromium.org has a significant advantage (the chromium version is 15KLOC as compared to the current version's 36K). On my machine the cs.chromium.org page loads in 4-5s, but it takes another 8-9s to attach the xrefs (not even all xrefs, it seems -- they appear to cut off around the anchor point on my link). The kythe-cs version takes about 12s to load on my machine, and this is even without significant performance tuning (with a few local tweaks to the page I was able to bring the total load time down to about 4s).

**Comment by kythe** [comment]
_**robinp** wrote at 2017-04-03 13:11:22_:

Thanks for the explanation Peter. The better resource utilization makes sense for the use case.

**Comment by kythe** [comment]
_**jrtom** wrote at 2017-10-18 17:10:21_:

Going through old tasks: what's the next step on this change? 
How will we decide whether this is something that we want to incorporate into the Kythe repo?

**Comment by kythe** [comment]
_**fromberger** wrote at 2018-05-21 20:39:39_:

As part of the move off Phabricator to GitHub pull requests, I'm abandoning this revision. If someone wants to pursue it further, please start a new issue on https://github.com/google/kythe/issues.

I took a clone of this patch, but I'm not sure it's actually in a buildable state at the moment. A lot of things have changed in Bazel since we last revisited this.

**Comment by kythe** [comment]
_**fromberger** abandoned revision at 2018-05-21 20:39:39_

**Comment by kythe** [comment]
_**pcc** wrote at 2018-05-21 20:46:33_:

Will the phabricator instance be going away as well? If so I will upload this change to github so that the code at least remains publicly accessible.

**Comment by kythe** [comment]
_**fromberger** wrote at 2018-05-21 20:51:39_:

>>! In (earlier comment), pcc wrote:
> Will the phabricator instance be going away as well? If so I will upload this change to github so that the code at least remains publicly accessible.

At some point, yes. I cloned the change locally but if you have a place to keep it on GitHub, that's probably even better.

**Comment by pcc** [comment]
I've pushed the code for this here: https://github.com/pcc/kythe/tree/kythecs-new
